### ISSUE TITLE
Inconsistent IAM Role Generation with API Gateway and Pure Lambdas

### ISSUE THREAD

**Issue by wollerman** [aws/chalice#685]
I have a chalice app that contains both API Gateway endpoints and pure lambda endpoints. When I run `chalice deploy` the first time everything runs smoothly and i see output such as:
```
Regen deployment package.
Creating role: my-service-dev
The following execution policy will be used: # followed by the policy i am using
Creating deployment package.
Creating lambda function: my-service-dev
Updating IAM policy for role: my-service-dev
Creating lambda function: my-service-dev-custom_pure_lambda
Initiating first time deployment.
Deploying to API Gateway stage: api
https://mylambda.execute-api.us-west-2.amazonaws.com/api/
```

However, if I make chances to my app and then again run `chalice deploy` I see a new role created for the pure lambda:

```
Regen deployment package.
Updating IAM policy for role: my-service-dev
Updating lambda function: my-service-dev
Updating IAM policy for role: my-service-dev
Regen deployment package.
Creating role: my-service-dev-custom_pure_lambda
The following execution policy will be used: # followed by the policy i am using
Updating lambda function: my-service-dev-custom_pure_lambda
API Gateway rest API already found: mylambda
Deploying to API Gateway stage: api
https://mylambda.execute-api.us-west-2.amazonaws.com/api/
```

I also do not see any reference to the roles in the `deployed.json` so I have to go look in the lamba console to find which role is being used.

This is a bug where roles are created at different times. I think either there should only be a single role or both roles should be created up front.

Let me know if I need to clarify anything. Thanks!

**Comment by stealthycoin** [comment]
Looks like a bug in the deployer.

**Comment by jamesls** [comment]
The new deployer should fix this (as of 1.2.0).  We've also added a number of test cases for configuring roles (https://github.com/aws/chalice/blob/master/tests/unit/deploy/test_deployer.py#L922) to ensure this continues to work properly.
